Doc/library/tkinter.rst

```diff
@@ -91,54 +91,27 @@ Python interpreter.
-
-Each :class:`Tk` object created by :mod:`tkinter` contains a Tcl interpreter.
-It also keeps track of which thread created that interpreter. Calls to
-:mod:`tkinter` can be made from any Python thread. Internally, if a call comes
-from a thread other than the one that created the :class:`Tk` object, an event
-is posted to the interpreter's event queue, and when executed, the result is
-returned to the calling Python thread.
-
-Tcl/Tk applications are normally event-driven, meaning that after initialization,
-the interpreter runs an event loop (i.e. :func:`Tk.mainloop`) and responds to events.
-Because it is single-threaded, event handlers must respond quickly, otherwise they
-will block other events from being processed. To avoid this, any long-running
-computations should not run in an event handler, but are either broken into smaller
-pieces using timers, or run in another thread. This is different from many GUI
-toolkits where the GUI runs in a completely separate thread from all application
-code including event handlers.
-
-If the Tcl interpreter is not running the event loop and processing events, any
-:mod:`tkinter` calls made from threads other than the one running the Tcl
-interpreter will fail.
-
-A number of special cases exist:
-
-  * Tcl/Tk libraries can be built so they are not thread-aware. In this case,
-    :mod:`tkinter` calls the library from the originating Python thread, even
-    if this is different than the thread that created the Tcl interpreter. A global
-    lock ensures only one call occurs at a time.
-
-  * While :mod:`tkinter` allows you to create more than one instance of a :class:`Tk`
-    object (with it's own interpreter), all interpreters that are part of the same
-    thread share a common event queue, which gets ugly fast. In practice, don't create
-    more than one instance of :class:`Tk` at a time. Otherwise, it's best to create
-    them in separate threads and ensure you're running a thread-aware Tcl/Tk build.
-
-  * Blocking event handlers are not the only way to prevent the Tcl interpreter from
-    reentering the event loop. It is even possible to run multiple nested event loops
-    or abandon the event loop entirely. If you're doing anything tricky when it comes
-    to events or threads, be aware of these possibilities.
-
-  * There are a few select :mod:`tkinter` functions that presently work only when
-    called from the thread that created the Tcl interpreter.
+Tkinter strives to allow any calls to its API from any Python threads, without any limits, as expected from a Python module. Due to Tcl's architectural restrictions, however, that stem from its vastly different threading model, this is not always possible.
+
+Tcl's execution model is based on cooperative multitasking. Control is passed between multiple interpreter instances by sending events (see `event-oriented programming -- Tcl/Tk wiki <https://wiki.tcl.tk/1772>`_ for details).
+
```

Comment:
I don't think the multiple interpreter thing is at all relevant. Main point.. one thread, event-oriented programming model. Calling Tk.mainloop causes Tcl to grab the next event, and process it via an event handler. Once that event handler finishes, the next event can be grabbed, etc. If for some reason that event handler takes a long time to run, there could be a long period of time before the next event is processed.

Reply:
To amplify this, in 99.99% of Tkinter apps, which would have one instance of Tkinter(), there's only ever going to be a single Tcl interpreter.

Reply:
Yes and no. Cooperative multitasking is _all_ about multiple independent agents coexisting in the same thread (in fact, for the purpose of cooperative multitasking, threads are irrelevant) and passing control between each other in some semi-predictable manner. Passing messages (which is what "events" effectively are) is just a very convenient way to do so that was popularized by Smalltalk back then AFAIK.

So, events are not just "signals that something happened", they _both_ pass information _and_ control.

In batch computations on a single-processor system, this theoretically gives the same benefits as multithreading, without the overhead. That's why no thread support in Tcl was seen as a no problem or even benefitial. (If the system needs to be responsive to external events that can arrive at any moment, this is not so "good" anymore... oh, well :) )

Multiple threads in Tcl were actually "slapped" on top of this machinery. Just like in a single thread, one instance sends a "message" to another and waits for the result. While that other one is checking for incoming messages this way or another. So, an inter-thread message still effectively leads to "passing control" between the two instances.

Reply:
I tried to explain all this in the text, and it invariably came across as irrelevant and confusing. THat's why I ultimately limited the text to mentioning just the most key trait of cooperatingg multitasking (so that the reader doesn't wonder how it is relevant) -- passing control.

Reply:
Considered "Information and control are passed...".
We need to somehow emphasize that they are passed together though, which this doesn't do.

Since "event" implies that it carries some data, I think the current formulation gives the most accurate impression in the end.